analysers/analyser_osmosis_powerline.py

```diff
@@ -24,28 +24,121 @@
     voltage
 FROM
-    ways
-    JOIN LATERAL (SELECT array_agg(lpad(v, 99, '0')) FROM unnest(regexp_split_to_array(ways.tags->'voltage','; *')) AS t(v)) AS t(voltage) ON TRUE
+    ways AS w
+    JOIN LATERAL (
+        SELECT array_agg(lpad(v, 99, '0'))
+        FROM (SELECT
+            unnest(array_cat(
+                array_fill(
+                    split_part(w.tags->'voltage', ';', 1)::text, -- voltage1 in voltage1;voltage2
+                    ARRAY[greatest(0, coalesce((w.tags->'circuits')::integer, 1) - (1 + length(coalesce(w.tags->'voltage', '')) - length(replace(coalesce(w.tags->'voltage',''), ';', ''))))]
+                ),
+                regexp_split_to_array(w.tags->'voltage', '; *'))
+            ) LIMIT coalesce((w.tags->'circuits')::integer, 1)
+        ) AS t(v)) AS t(voltage)
+        ON TRUE
 WHERE
-    ways.tags != ''::hstore AND
-    ways.tags?'power' AND
-    ways.tags->'power' IN ('line', 'minor_line', 'cable')
+    w.tags != ''::hstore AND
+    w.tags?'power' AND
+    w.tags->'power' IN ('line', 'minor_line', 'cable') AND
+    w.tags?'voltage' AND
+    (
+        NOT w.tags?'circuits' OR
+        w.tags->'circuits' ~ '^[0-9]+$'
+    )
+
+UNION ALL
+
+SELECT
+    w.id AS wid,
+    unnest('{NULL}' || w.nodes[1:array_length(w.nodes, 1) - 1]) AS nid_prec,
+    unnest(w.nodes) AS nid,
+    unnest(w.nodes[2:array_length(w.nodes, 1)]) AS nid_next,
+    w.tags->'cables' AS cables,
+    coalesce((w.tags->'circuits')::integer, 1) AS circuits,
+    coalesce(w.tags->'location', 'overhead') AS location,
+    NULL AS voltage
+FROM
+   ways AS w
+WHERE
+    w.tags != ''::hstore AND
+    w.tags?'power' AND
+    w.tags->'power' IN ('line', 'minor_line', 'cable') AND
+    (
+        NOT w.tags?'voltage' OR (
+            w.tags?'circuits' AND
+            NOT(w.tags->'circuits' ~ '^[0-9]+$')
+        )
+    )
+"""
+
+# Build junctions knowledge
+# Topoedges are couples attached to a given node with their neighbors.
+# Topoedges are agregated by nodes and location (two *overhead* lines between two node will give a single topoedge)
+# Involved nodes are not necessary power, particularly on cables
+sql02 = """
+CREATE TEMP TABLE power_lines_topoedges AS
+WITH topotuples as (
+    SELECT
+        n.wid,
+        n.nid # n.nid_prec AS tid,
+        n.nid,
+        n.location,
+        n.cables,
+        n.circuits,
+        voltage
+    FROM
+        power_lines_nodes AS n
+    WHERE
+        nid_prec IS NOT NULL
+
+    UNION ALL
+
+    SELECT
+        n.wid,
+        n.nid # n.nid_next as tid,
+        n.nid,
+        n.location,
+        n.cables,
+        n.circuits,
+        voltage
+    FROM
+        power_lines_nodes AS n
+    WHERE
+        nid_next IS NOT NULL
+)
+
+SELECT
+    p.nid,
+    p.tid,
+    p.location,
+    count(distinct p.wid) AS cw,
+    sum(p.circuits::integer) AS circuits,
+    regexp_split_to_array(string_agg(array_to_string(p.voltage, ';'), ';'), '; *') AS voltage
```

Comment:
```suggestion
    regexp_split_to_array(string_agg(array_to_string(p.voltage, ';'), ';'), '; *') AS voltages
```

Reply:
Renaming as such should be done on `power_lines_nodes` and `power_lines_topoedges` tables as well, right?

Reply:
Yes